### PR TITLE
[WIP][IMP] base: control center to check access rules and right of each user / model

### PR DESCRIPTION
**Pull request by fmdl**
**Description of the issue/feature this PR addresses:**
With the `ir.rule` it is hard to check if everything is good.

**To check by user and model :**
Go to Technical/Model
You see the number of user who have access by ACL.
image

If you click on this button you can see for each mode (read, ...) the number of records who the user have access.
image

You can click on the Read for exemple to see the records.

Note : you need to use with uid == 1

Limitation : you can check only the rules with the current user company.

@odony what do you think ?

Nota : could you work on a `sudo(company_id=xx)` to by pass the current user company ?

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: odoo/addons/base/views/res_users_views.xml [hunk 1]
```diff
@@ -236,6 +236,15 @@
                     <field name="login"/>
                     <field name="lang"/>
                     <field name="login_date"/>
+                    <field name="company_id"/>
```
Comment (mreficent) [inline odoo/addons/base/views/res_users_views.xml]: maybe could you add `groups="base.group_multi_company"` here

File: odoo/addons/base/models/res_users.py [hunk 2]
```diff
@@ -279,6 +284,72 @@ def init(self):
...
+        return self._action_open_record('create')
+
+    def action_open_unlink_record(self):
+        return self._action_open_record('unlink')
+
+    def _action_open_record(self, mode):
+        self.ensure_one()
+        if self.env.uid != SUPERUSER_ID:
+            return UserError(_('Only Super Admin can perform this action.'))
+        if self.env.user.company_id != self.company_id:
```
Comment (mreficent) [inline odoo/addons/base/models/res_users.py]:
```python
if self.company_id not in self.env.user.company_ids:
```
please

Comment (mreficent) [reply odoo/addons/base/models/res_users.py]: or maybe
```python
if self.company_id not in (self.env.user.company | self.env.user.company.child_ids).ids:
```

**Comment by fmdl** [comment]
@mreficent I will fix, what do you think about this improvement ?

**Comment by mreficent** [comment]
@fmdl very good!

**Comment by fmdl** [comment]
Thanks, I wait the answer of @odony to continue the dev.

**Comment by fmdl** [comment]
@alexis-via

**Comment by fmdl** [comment]
@odony what do you think ?

**Comment by fmdl** [comment]
cc @Yenthe666

**Comment by Yenthe666** [comment]
@fmdl you know that there is something very similar on users right now in master right?
image

It might coexist though as this is another easy way to check on a model level and not user level. Have you checked the performance difference for those computes on `res.partner`? I assume they'd have effect on multiple places throughout the system.

**Comment by fmdl** [comment]
Yes I know (it is my idea https://github.com/odoo/odoo/pull/23452, and there is an open PR to fix a bug https://github.com/odoo/odoo/pull/28488). But with this you cannot check if the ir.rules works correctly.

about the perf, the fields is only computed if you come from the model https://github.com/odoo/odoo/pull/29122/files#diff-c087b09b8455bbf3fa546a56c6e271bcR333 and https://github.com/odoo/odoo/pull/29122/files#diff-bb1e3d0f2e95b26bcd024950824e022fR129

**Comment by Yenthe666** [comment]
Okay I missed that. This looks good. @odony could you please give this PR some attention?

**Comment by fmdl** [comment]
ping @odony

**Comment by odony** [comment]
I don't think I understanf the point of seeing how many records does User X sees of Model Y?  In a real company those numbers are big and change all the time. Who knows how many records each user is supposed to see and whether this number is correct based on the rules?

It would take a long investigation to verify this number on an active database, versus just looking at the ACLs and group memberships and verifying them. If the number turns out to be incorrect, you'll end up verifying the ACLs anyway, no?

And what does 0 mean? No access or no data?

I think a simple link to the list of users who have access could be useful, but even that usefulness is limited because you need to know through which group they have access.

Basically the info is all there in the ACLs and Rules tabs, in a very straighforward way.

Feel free to discuss with @tde-banana-odoo (who self-assigned it), but for mee it seems we're increasing the bloat already created by #25587.

**Comment by tde-banana-odoo** [comment]
Hello,

Not sure to have all the history in mind, but globally the idea was to have shortcuts to rules applied to a given user directly from its user form. I don't think the count itself adds any information; main purpose was to have stat buttons acting like shortcuts from a user to acls and rules.

If you have an issue with a user that cannot access to a given model (or seems to be able to access a model but he shouldn't), you will go to its acls, filter on a model and look at them. Knowing user Tde can read 17 models is not an information we can use on a daily basis, unless I am wrong. However if Tde cannot create tasks anymore, I can jump from its user profile to the acls and see what apply in the list view while filtering on task, project, ...

In current master it seems there is a counter that you want to tweak; according to me counter itself could be removed but we should ensure displayed information is relevant to the user from which we came.

Did I miss something ?

Thanks,

**Comment by fmdl** [comment]
@tde-banana-odoo this POC is maybe not the best solution to make an audit of ACL and rules for one user or to test rules (without using the user account to test).
